src/btree/bt_handle.c

```diff
@@ -854,15 +875,16 @@ __btree_get_last_recno(WT_SESSION_IMPL *session)
     }
 
     /*
-     * The endpoint for append is global; read the last page with global visibility to make sure
-     * that if the end of the tree is truncated we don't start appending in the truncated space
-     * unless the truncation has become globally visible. (Note that this path does not examine the
-     * visibility of individual data items; it only checks whether whole pages are deleted.)
+     * The endpoint for append is global; read the last page with global visibility (even if it's
+     * deleted) to make sure that if the end of the tree is truncated, the tree walk finds the
+     * correct page. (Note that this path does not examine the visibility of individual data items;
+     * it only checks whether whole pages are deleted.)
      */
-    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL;
+    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL | WT_READ_SEE_DELETED;
 
     next_walk = NULL;
-    WT_RET(__wt_tree_walk(session, &next_walk, flags));
+    WT_RET(
+      __wt_tree_walk_custom_skip(session, &next_walk, &__btree_last_recno_no_skip, NULL, flags));
```

Comment:
Why does the walk need both a custom skip implementation and a new flag. I'd have thought that just the flag would be enough?

Reply:
Oops, thanks - I mis-read the interaction in `__tree_walk_internal` between the callback and the `WT_READ_SKIP_DELETED` flag. Fixed in b1956bd.